File: ios/MegaPDF/BusyState.swift
import Foundation
import UIKit

// Feedback while the app is working (#145): one busy state per open document.
//
//   * The moment an operation starts it is recorded here, so the control that started it
//     is disabled and repeat taps are ignored (`begin` answers nil while blocking work runs).
//   * An indeterminate indicator appears only once the work has run for 0.5 s, so quick
//     work never flickers, and once shown it stays at least 0.3 s.
//   * Document-level work (opening, saving, search) shows a strip under the top toolbar;
//     page-level work (the page checks, applying a change) a small spinner on the page or
//     the line. Each has its own indicator and timing.

/// What the indicator says. Every label is also what VoiceOver reads.
enum BusyLabel: Equatable {
    case opening
    case saving
    case checkingSavedFile
    case checkingPage
    case applying
    case searching
    /// A Markdown export (#386) -- distinct from `.saving`, since it isn't one (see
    /// `ViewerModel.exportMarkdownFile`).
    case exportingText

    var text: String {
        switch self {
        case .opening: return String(localized: "Opening…")
        case .saving: return String(localized: "Saving…")
        case .checkingSavedFile: return String(localized: "Checking the saved file…")
        case .checkingPage: return String(localized: "Checking this page…")
        case .applying: return String(localized: "Applying…")
        case .searching: return String(localized: "Searching…")
        case .exportingText: return String(localized: "Exporting…")
        }
    }
}

/// Where the indicator for a piece of work goes.
enum BusyScope: Equatable {
    /// The strip under the navigation bar.
    case document
    /// A spinner on the page, centred on `rect` (PDF points) when one is given.
    case page(Int, PdfRect?)

    var isDocument: Bool { self == .document }
}

/// One piece of running work.
struct BusyWork: Identifiable, Equatable {
    let id: Int
    var label: BusyLabel
    var scope: BusyScope
    /// Blocking work stops editing and further blocking work until it ends. Search doesn't.
    let blocking: Bool
    /// Close, Discard and the file commands wait for it too (saves, password changes, applying).
    var blocksFileCommands: Bool
    /// Off while the work waits on the person (the #139 warning is up): no spinner behind an alert.
    var showsIndicator: Bool
}

/// A handle on running work, returned by `begin`.
struct BusyToken: Equatable {
    fileprivate let id: Int
}

/// The clock the indicator timing runs on; tests drive a manual one.
protocol BusyScheduler {
    var now: TimeInterval { get }
    /// Runs `action` on the main actor after `delay` seconds, unless the returned handle is cancelled first.
    func schedule(after delay: TimeInterval, _ action: @escaping @MainActor () -> Void) -> BusyScheduled
}

protocol BusyScheduled {
    func cancel()
}

/// The real clock.
struct SystemBusyScheduler: BusyScheduler {
    var now: TimeInterval { ProcessInfo.processInfo.systemUptime }

    func schedule(after delay: TimeInterval, _ action: @escaping @MainActor () -> Void) -> BusyScheduled {
        let item = DispatchWorkItem { MainActor.assumeIsolated { action() } }
        DispatchQueue.main.asyncAfter(deadline: .now() + max(0, delay), execute: item)
        return item
    }
}

extension DispatchWorkItem: BusyScheduled {}

@MainActor
final class BusyState: ObservableObject {
    nonisolated static let showDelay: TimeInterval = 0.5
    nonisolated static let minimumVisible: TimeInterval = 0.3

    /// Everything running, oldest first. Changes the instant work begins or ends.
    @Published private(set) var works: [BusyWork] = []
    /// The strip under the toolbar, when it is showing.
    @Published private(set) var strip: BusyWork?
    /// The spinner on a page, when it is showing.
    @Published private(set) var pageIndicator: BusyWork?

    private let scheduler: BusyScheduler
    private let showDelay: TimeInterval
    private let minimumVisible: TimeInterval
    private let announce: @MainActor (String) -> Void
    private var nextId = 1
    private lazy var stripTiming = IndicatorTiming(owner: self, isDocument: true)
    private lazy var pageTiming = IndicatorTiming(owner: self, isDocument: false)

    init(scheduler: BusyScheduler = SystemBusyScheduler(),
         showDelay: TimeInterval = BusyState.showDelay,
         minimumVisible: TimeInterval = BusyState.minimumVisible,
         announce: @escaping @MainActor (String) -> Void = { UIAccessibility.post(notification: .announcement, argument: $0) }) {
        self.scheduler = scheduler
        self.showDelay = showDelay
        self.minimumVisible = minimumVisible
        self.announce = announce
    }

    /// Blocking work is running: page taps, tools and edits wait.
    var isBlocked: Bool { works.contains { $0.blocking } }

    /// Close, Discard and file commands wait: a save, a password change or a change being applied runs.
    var blocksFileCommands: Bool { works.contains { $0.blocksFileCommands } }

    /// Starts work. Blocking work answers nil while other blocking work runs: the tap is ignored.
    func begin(_ label: BusyLabel, scope: BusyScope, blocking: Bool = true,
               blocksFileCommands: Bool = false, showsIndicator: Bool = true) -> BusyToken? {
        if blocking && isBlocked { return nil }
        let work = BusyWork(id: nextId, label: label, scope: scope, blocking: blocking,
                            blocksFileCommands: blocksFileCommands, showsIndicator: showsIndicator)
        nextId += 1
        works.append(work)
        refresh()
        return BusyToken(id: work.id)
    }

    /// Changes what running work says, where it shows, or whether it shows at all.
    func update(_ token: BusyToken, label: BusyLabel? = nil, scope: BusyScope? = nil,
                blocksFileCommands: Bool? = nil, showsIndicator: Bool? = nil) {
        guard let i = works.firstIndex(where: { $0.id == token.id }) else { return }
        var work = works[i]
        if let label { work.label = label }
        if let scope { work.scope = scope }
        if let blocksFileCommands { work.blocksFileCommands = blocksFileCommands }
        if let showsIndicator { work.showsIndicator = showsIndicator }
        guard work != works[i] else { return }
        works[i] = work
        refresh()
    }

    /// Ends work. Safe to call twice.
    func end(_ token: BusyToken) {
        guard let i = works.firstIndex(where: { $0.id == token.id }) else { return }
        works.remove(at: i)
        refresh()
    }

    /// Runs `body` as busy work, ending it however `body` finishes. Nil when blocked.
    func run<T>(_ label: BusyLabel, scope: BusyScope, blocking: Bool = true, blocksFileCommands: Bool = false,
                _ body: (BusyToken) async throws -> T) async rethrows -> T? {
        guard let token = begin(label, scope: scope, blocking: blocking,
                                blocksFileCommands: blocksFileCommands) else { return nil }
        defer { end(token) }
        return try await body(token)
    }

    /// Drops everything at once, indicators included.
    func reset() {
        works = []
        stripTiming.reset()
        pageTiming.reset()
        strip = nil
        pageIndicator = nil
    }

    /// The document closed: its page-level work is over, and the spinner goes at once. Document-level
    /// work stays, because whoever started it (an open, a password change that reopens) ends it.
    func endPageWork() {
        works.removeAll { !$0.scope.isDocument }
        pageTiming.reset()
        pageIndicator = nil
        refresh()
    }

    // MARK: - indicator timing

    /// What each indicator should show now: the newest blocking work, else the newest other work.
    fileprivate func candidate(document: Bool) -> BusyWork? {
        let eligible = works.filter { $0.showsIndicator && $0.scope.isDocument == document }
        return eligible.last(where: { $0.blocking }) ?? eligible.last
    }

    fileprivate var clockNow: TimeInterval { scheduler.now }

    fileprivate func schedule(after delay: TimeInterval, _ action: @escaping @MainActor () -> Void) -> BusyScheduled {
        scheduler.schedule(after: delay, action)
    }

    fileprivate var delays: (show: TimeInterval, minimum: TimeInterval) { (showDelay, minimumVisible) }

    fileprivate func publish(_ work: BusyWork?, document: Bool) {
        if document {
            let appearing = strip == nil && work != nil
            if strip != work { strip = work }
            if appearing, let work { announce(work.label.text) }
        } else if pageIndicator != work {
            pageIndicator = work
        }
    }

    private func refresh() {
        stripTiming.refresh()
        pageTiming.refresh()
    }
}

/// The 0.5 s / 0.3 s rule for one indicator.
@MainActor
private final class IndicatorTiming {
    private unowned let owner: BusyState
    private let isDocument: Bool
    private var showTimer: BusyScheduled?
    private var hideTimer: BusyScheduled?
    private var shownAt: TimeInterval?

    init(owner: BusyState, isDocument: Bool) {
        self.owner = owner
        self.isDocument = isDocument
    }

    func refresh() {
        let candidate = owner.candidate(document: isDocument)
        if let candidate {
            hideTimer?.cancel()
            hideTimer = nil
            if shownAt != nil {
                owner.publish(candidate, document: isDocument)
            } else if showTimer == nil {
                // The delay runs from when there was first something to show.
                showTimer = owner.schedule(after: owner.delays.show) { [weak self] in self?.show() }
            }
        } else {
            showTimer?.cancel()
            showTimer = nil
            guard let shownAt, hideTimer == nil else { return }
            let remaining = owner.delays.minimum - (owner.clockNow - shownAt)
            if remaining <= 0 {
                hide()
            } else {
                hideTimer = owner.schedule(after: remaining) { [weak self] in self?.hide() }
            }
        }
    }

    func reset() {
        showTimer?.cancel()
        hideTimer?.cancel()
        showTimer = nil
        hideTimer = nil
        shownAt = nil
    }

    private func show() {
        showTimer = nil
        guard let candidate = owner.candidate(document: isDocument) else { return }
        shownAt = owner.clockNow
        owner.publish(candidate, document: isDocument)
    }

    private func hide() {
        hideTimer = nil
        // Work that began while the indicator lingered keeps it up.
        if let candidate = owner.candidate(document: isDocument) {
            owner.publish(candidate, document: isDocument)
            return
        }
        shownAt = nil
        owner.publish(nil, document: isDocument)
    }
}

File: android/app/src/main/java/com/megapdf/android/BusyState.kt
package com.megapdf.android

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.megapdf.engine.PdfRect
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/** What the app is busy with (#145); each has its label in the string catalogue. */
enum class BusyLabel(val stringId: Int) {
    OPENING(R.string.busy_opening),
    SAVING(R.string.saving),
    VERIFYING_SAVE(R.string.busy_verifying_save),
    SEARCHING(R.string.busy_searching),
    CHECKING_PAGE(R.string.busy_checking_page),
    APPLYING(R.string.busy_applying),
    /** Exporting the document's text as Markdown (#386) — its own label, not [SAVING]: a
     * Markdown export is not a save of the document (see [ViewerViewModel.exportMarkdown]). */
    EXPORTING(R.string.busy_exporting),
}

/** Where page-level work shows its spinner: a page, and the line on it when there is one. */
data class BusySpot(val pageIndex: Int, val rect: PdfRect? = null)

/**
 * One busy indicator (#145): the strip under the top bar, or the spinner on a page.
 *
 * [isActive] turns on the moment work begins, so the control used can be disabled and repeat
 * taps ignored at once. [isVisible] turns on only once work has run for [showAfterMs], so quick
 * work never flickers an indicator, and once on it stays at least [minShownMs], so it never
 * blinks. While several pieces of work overlap, the latest one's [label] and [spot] show.
 */
class BusyIndicator internal constructor(
    private val scope: CoroutineScope,
    private val now: () -> Long,
    private val showAfterMs: Long,
    private val minShownMs: Long,
) {
    private val active = ArrayList<BusyToken>()
    private var showJob: Job? = null
    private var hideJob: Job? = null
    private var shownAt = 0L

    var isActive: Boolean by mutableStateOf(false)
        private set
    var isVisible: Boolean by mutableStateOf(false)
        private set
    var label: BusyLabel? by mutableStateOf(null)
        private set
    var spot: BusySpot? by mutableStateOf(null)
        private set

    internal fun begin(label: BusyLabel, spot: BusySpot?, onEnd: () -> Unit): BusyToken {
        val token = BusyToken(this, label, spot, onEnd)
        active += token
        update()
        return token
    }

    internal fun changed() = update()

    internal fun ended(token: BusyToken) {
        active.remove(token)
        update()
    }

    /** Drops every piece of work at once, without waiting out the minimum: its document is gone. */
    internal fun reset() {
        active.forEach { it.detach() }
        active.clear()
        showJob?.cancel()
        showJob = null
        hideJob?.cancel()
        hideJob = null
        isActive = false
        isVisible = false
        label = null
        spot = null
    }

    private fun update() {
        val latest = active.lastOrNull()
        if (latest != null) {
            isActive = true
            label = latest.label
            spot = latest.spot
            hideJob?.cancel()
            hideJob = null
            if (!isVisible && showJob == null) {
                showJob = scope.launch {
                    delay(showAfterMs)
                    showJob = null
                    isVisible = true
                    shownAt = now()
                }
            }
            return
        }
        isActive = false
        showJob?.cancel()
        showJob = null
        if (!isVisible) {
            label = null
            spot = null
        } else if (hideJob == null) {
            val remaining = minShownMs - (now() - shownAt)
            if (remaining <= 0) {
                hide()
            } else {
                hideJob = scope.launch {
                    delay(remaining)
                    hideJob = null
                    hide()
                }
            }
        }
    }

    private fun hide() {
        if (active.isNotEmpty()) return
        isVisible = false
        label = null
        spot = null
    }
}

/** One piece of busy work; [end] it in a `finally`. Ending twice, or after a reset, does nothing. */
class BusyToken internal constructor(
    private var owner: BusyIndicator?,
    label: BusyLabel,
    spot: BusySpot?,
    private var onEnd: (() -> Unit)?,
) {
    var label: BusyLabel = label
        private set
    var spot: BusySpot? = spot
        private set

    /** The same work moving on to its next step, e.g. Saving… to Checking the saved file…. */
    fun relabel(label: BusyLabel, spot: BusySpot? = this.spot) {
        this.label = label
        this.spot = spot
        owner?.changed()
    }

    fun end() {
        val indicator = owner ?: return
        owner = null
        onEnd?.invoke()
        onEnd = null
        indicator.ended(this)
    }

    internal fun detach() {
        owner = null
        onEnd = null
    }
}

/**
 * The busy state of the open document (#145): one strip for document-level work (open, save,
 * search) and one spinner for page-level work (the page check, the text-edit check, applying a
 * change). The view model resets it whenever a document opens or closes.
 *
 * Call it from one thread (the main thread in the app). [now] and the delays are injectable so
 * the timing is tested on a virtual clock.
 */
class BusyState(
    scope: CoroutineScope,
    now: () -> Long = { System.nanoTime() / 1_000_000 },
    showAfterMs: Long = SHOW_AFTER_MS,
    minShownMs: Long = MIN_SHOWN_MS,
) {
    /** The strip under the top app bar. */
    val document = BusyIndicator(scope, now, showAfterMs, minShownMs)

    /** The small spinner on a page or a line. */
    val page = BusyIndicator(scope, now, showAfterMs, minShownMs)

    private var locks = 0

    /**
     * True while a save or a password change runs: editing, Close and Back, and the file
     * commands are disabled until it ends.
     */
    var locksDocument: Boolean by mutableStateOf(false)
        private set

    /** Document-level work; [locks] for a save or a password change. */
    fun beginDocument(label: BusyLabel, locks: Boolean = false): BusyToken {
        if (locks) {
            this.locks++
            locksDocument = true
        }
        return document.begin(label, null) {
            if (locks) {
                this.locks--
                locksDocument = this.locks > 0
            }
        }
    }

    /** Page-level work, shown at [spot]. */
    fun beginPage(label: BusyLabel, spot: BusySpot?): BusyToken = page.begin(label, spot) {}

    /** Page-level [block], with its spinner. */
    suspend fun <T> pageWork(label: BusyLabel, spot: BusySpot?, block: suspend () -> T): T {
        val token = beginPage(label, spot)
        try {
            return block()
        } finally {
            token.end()
        }
    }

    /** Forgets all work: another document is opening, or this one closed. */
    fun reset() {
        document.reset()
        page.reset()
        locks = 0
        locksDocument = false
    }

    companion object {
        /** Work quicker than this shows no indicator at all. */
        const val SHOW_AFTER_MS = 500L

        /** An indicator that appeared stays at least this long. */
        const val MIN_SHOWN_MS = 300L
    }
}
